packages/run-protocol/src/runStake/attestation.js

```diff
@@ -0,0 +1,256 @@
+  Attestation: 'Attestation',
+});
+
+/**
+ * Find-or-create value in store.
+ *
+ * @type {<K, V>(store: Store<K, V>, key: K, thunk: () => V) => V}
+ */
+const getOrElse = (store, key, make) => {
+  if (store.has(key)) {
+    return store.get(key);
+  }
+  const value = make();
+  store.init(key, value);
+  return value;
+};
+
+/**
+ * To support `makeAttestation`, we directly mint attestation payments.
+ * We still use a ZCFMint because returning attestations is done
+ * through Zoe offers.
+ *
+ * @param {ZCF} zcf
+ * @param {ZCFMint} zcfMint
+ * @param {Amount} amountToMint
+ * @returns {Promise<Payment>}
+ */
+const mintZCFMintPayment = (zcf, zcfMint, amountToMint) => {
+  const { userSeat, zcfSeat } = zcf.makeEmptySeatKit();
+  zcfMint.mintGains(harden({ [KW.Attestation]: amountToMint }), zcfSeat);
+  zcfSeat.exit();
+  return E(userSeat).getPayout(KW.Attestation);
+};
+
+/**
+ * Make an ZCFMint attentuated so that minting represents
+ * putting a lien on staked assets. Payments can be returned,
+ * releasing (part of) the lien.
+ *
+ * @param {ZCF} zcf
+ * @param {string} keyword for use in makeZCFMint
+ * @param {Brand<'nat'>} stakeBrand brand of the staked assets
+ * @param {ERef<StakingAuthority>} lienBridge bridge to account state
+ */
+const makeAttestationKit = async (zcf, keyword, stakeBrand, lienBridge) => {
+  const { add, subtract, makeEmpty, isGTE, coerce } = AmountMath;
+  const empty = makeEmpty(stakeBrand);
+  const zcfMint = await zcf.makeZCFMint(keyword, AssetKind.COPY_BAG);
+  const { issuer, brand: attBrand } = zcfMint.getIssuerRecord();
+
+  /**
+   * @param {string} address
+   * @param {Amount<'nat'>} lienedAmount
+   * @returns {Amount<'copyBag'>}
+   */
+  const wrapLienedAmount = (address, lienedAmount) =>
+    AmountMath.make(attBrand, makeCopyBag([[address, lienedAmount.value]]));
+
+  /** @param {Amount<'copyBag'>} attestationAmount */
+  const unwrapLienedAmount = attestationAmount => {
+    assert(
+      attestationAmount.brand === attBrand,
+      X`The escrowed attestation ${attestationAmount} was not of the attestation brand ${attBrand}`,
+    );
+    // Caller is responsible to split payload of >1 item.
+    fit(attestationAmount.value.payload, harden([[M.string(), M.bigint()]]));
+    /** @type {[string, bigint][]} */
+    const [[address, valueReturned]] = attestationAmount.value.payload;
+    const lienedAmount = AmountMath.make(stakeBrand, valueReturned);
+    return { address, lienedAmount };
+  };
+
+  /** @type {(hi: Amount, lo: Amount) => Amount} */
+  const subtractOr0 = (hi, lo) => (isGTE(hi, lo) ? subtract(hi, lo) : empty);
+  // check that x <= hi - lo
+  const checkDelta = (x, label, hi, lo) => {
+    const bound = subtractOr0(hi, lo);
+    assert(
+      isGTE(bound, coerce(stakeBrand, x)),
+      // TODO: X`...${raw(label)}...`
+      X`Only ${bound} was ${label}, but an attestation was attempted for ${x}`,
+    );
+  };
+  const checkAccountState = async (address, toLien) => {
```

Comment:
do we have a convention for what `checkFoo` means? I've seen it used to throw and to return a value. I think returning a value is bad and throwing is good. But if we are going to return values from `checkFoo` then I think we should have a clear verb for throwing, like `validateFoo`.

Reply:
There's a convention in `@agoric/store` to use `isFoo(x)` to return a boolean about whether `x` is a `Foo` and `assertFoo(x)` to throw if `x` is not. Both are implemented in terms of `checkFoo(x, checker)` where `checker(cond, detail)` is a `cond => cond` for `isFoo` and `assert` for `assertFoo`.

So I changed this to `assertAccountStateOK`.